### PR TITLE
fix: stop memory leak from orphaned CR reflector goroutines on repeated CRD discovery

### LINKED ISSUE #2867: Elevated Memory Utilization (v2.18.0)
**What happened**:
Memory on v2.18.0 is elevated, more than double what it used at v2.17.0 at peak utilization. In my configuration, it never once was OOMKilled at v2.17.0 with 512MiB limit, where on v2.18.0 it will be OOMKilled at 1GiB.

**What you expected to happen**:
Memory utilization to be consistent between versions

**How to reproduce it (as minimally and precisely as possible)**:
Deployment Spec:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "14"
    meta.helm.sh/release-name: victoria-metrics
    meta.helm.sh/release-namespace: monitoring
  labels:
    app.kubernetes.io/component: metrics
    app.kubernetes.io/instance: victoria-metrics
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/part-of: kube-state-metrics
    app.kubernetes.io/version: 2.18.0
    helm.sh/chart: kube-state-metrics-6.3.0
    helm.toolkit.fluxcd.io/name: victoria-metrics
    helm.toolkit.fluxcd.io/namespace: monitoring
  name: victoria-metrics-kube-state-metrics
  namespace: monitoring
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: victoria-metrics
      app.kubernetes.io/name: kube-state-metrics
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      labels:
        app.kubernetes.io/component: metrics
        app.kubernetes.io/instance: victoria-metrics
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: kube-state-metrics
        app.kubernetes.io/part-of: kube-state-metrics
        app.kubernetes.io/version: 2.18.0
        helm.sh/chart: kube-state-metrics-6.3.0
    spec:
      automountServiceAccountToken: true
      containers:
      - args:
        - --port=8080
        - --resources=certificatesigningrequests,configmaps,cronjobs,daemonsets,deployments,endpoints,horizontalpodautoscalers,ingresses,jobs,leases,limitranges,mutatingwebhookconfigurations,namespaces,networkpolicies,nodes,persistentvolumeclaims,persistentvolumes,poddisruptionbudgets,pods,replicasets,replicationcontrollers,resourcequotas,secrets,services,statefulsets,storageclasses,validatingwebhookconfigurations,volumeattachments
        - --custom-resource-state-config-file=/etc/customresourcestate/config.yaml
        image: registry.k8s.io/kube-state-metrics/kube-state-metrics:v2.18.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /livez
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: kube-state-metrics
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /readyz
            port: 8081
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        resizePolicy:
        - resourceName: memory
          restartPolicy: NotRequired
        - resourceName: cpu
          restartPolicy: NotRequired
        resources:
          limits:
            memory: 512Mi
          requests:
            cpu: 25m
            memory: 256Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/customresourcestate
          name: customresourcestate-config
          readOnly: true
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: harbor-pull-secret
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 65534
        runAsGroup: 65534
        runAsNonRoot: true
        runAsUser: 65534
        seccompProfile:
          type: RuntimeDefault
      serviceAccount: victoria-metrics-kube-state-metrics
      serviceAccountName: victoria-metrics-kube-state-metrics
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          name: victoria-metrics-kube-state-metrics-customresourcestate-config
        name: customresourcestate-config
```

Custom Resource State ConfigMap (used for VPA metrics)
```yaml
apiVersion: v1
data:
  config.yaml: |
    kind: CustomResourceStateMetrics
    spec:
      resources:
      - groupVersionKind:
          group: autoscaling.k8s.io
          kind: VerticalPodAutoscaler
          version: v1
        labelsFromPath:
          namespace:
          - metadata
          - namespace
          target_api_version:
          - spec
          - targetRef
          - apiVersion
          target_kind:
          - spec
          - targetRef
          - kind
          target_name:
          - spec
          - targetRef
          - name
          verticalpodautoscaler:
          - metadata
          - name
        metrics:
        - each:
            info:
              labelsFromPath:
                name:
                - metadata
                - name
            type: Info
          help: VPA container recommendations. Kubernetes labels converted to Prometheus
            labels
          name: verticalpodautoscaler_labels
        - commonLabels:
            resource: memory
            unit: byte
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - target
              - memory
            type: Gauge
          help: VPA container recommendations for memory. Target resources the VerticalPodAutoscaler
            recommends for the container.
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_target
        - commonLabels:
            resource: memory
            unit: byte
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - lowerBound
              - memory
            type: Gauge
          help: VPA container recommendations for memory. Minimum resources the container
            can use before the VerticalPodAutoscaler updater evicts it
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_lowerbound
        - commonLabels:
            resource: memory
            unit: byte
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - upperBound
              - memory
            type: Gauge
          help: VPA container recommendations for memory. Maximum resources the container
            can use before the VerticalPodAutoscaler updater evicts it
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_upperbound
        - commonLabels:
            resource: memory
            unit: byte
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - uncappedTarget
              - memory
            type: Gauge
          help: VPA container recommendations for memory. Target resources the VerticalPodAutoscaler
            recommends for the container ignoring bounds
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_uncappedtarget
        - commonLabels:
            resource: cpu
            unit: core
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - target
              - cpu
            type: Gauge
          help: VPA container recommendations for cpu. Target resources the VerticalPodAutoscaler
            recommends for the container.
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_target
        - commonLabels:
            resource: cpu
            unit: core
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - lowerBound
              - cpu
            type: Gauge
          help: VPA container recommendations for cpu. Minimum resources the container
            can use before the VerticalPodAutoscaler updater evicts it
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_lowerbound
        - commonLabels:
            resource: cpu
            unit: core
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - upperBound
              - cpu
            type: Gauge
          help: VPA container recommendations for cpu. Maximum resources the container
            can use before the VerticalPodAutoscaler updater evicts it
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_upperbound
        - commonLabels:
            resource: cpu
            unit: core
          each:
            gauge:
              labelsFromPath:
                container:
                - containerName
              path:
              - status
              - recommendation
              - containerRecommendations
              valueFrom:
              - uncappedTarget
              - cpu
            type: Gauge
          help: VPA container recommendations for cpu. Target resources the VerticalPodAutoscaler
            recommends for the container ignoring bounds
          name: verticalpodautoscaler_status_recommendation_containerrecommendations_uncappedtarget
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: victoria-metrics
    meta.helm.sh/release-namespace: monitoring
  labels:
    app.kubernetes.io/component: metrics
    app.kubernetes.io/instance: victoria-metrics
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/part-of: kube-state-metrics
    app.kubernetes.io/version: 2.18.0
    helm.sh/chart: kube-state-metrics-6.3.0
    helm.toolkit.fluxcd.io/name: victoria-metrics
    helm.toolkit.fluxcd.io/namespace: monitoring
  name: victoria-metrics-kube-state-metrics-customresourcestate-config
  namespace: monitoring
```

**Anything else we need to know?**:
It doesn't immediately OOM, the memory grows over time, suggesting a memory leak. 

Here's a memory usage graph of the 12 hours I had version 2.18.0 deployed.

Image

**Environment**:

* kube-state-metrics version: 2.18.0
* Kubernetes version (use `kubectl version`): 1.35.0
* Cloud provider or hardware configuration: Self-hosted, kubeadm cluster
* Other info:

### PR DESCRIPTION
**Pull request by bhope**
Elevated and unbounded memory growth introduced in v2.18.0 when custom resource state config is in use.

Root Causes

1. AppendToMap overwrites stop channels and appends duplicate kinds on every call (internal/discovery/types.go). Since PollForCacheUpdates calls it for every known GVK each cycle, old stop channels were silently replaced, orphaning any reflector goroutine blocking on them.
2. CR reflectors ignore context cancellation (internal/store/builder.go). Unlike standard reflectors started with reflector.Run(b.ctx.Done()), custom resource reflectors were started with only their GVK-specific stop channel - no context cancellation path at all.

Fix
- AppendToMap: skip the append if the kind already exists; skip make(chan struct{}) if a channel already exists for the GVK.
- startReflector: wrap the GVK stop channel with a bridge goroutine that also selects on b.ctx.Done(), so CR reflectors stop on both CRD deletion and context cancellation.

Also, added tests to cover idempotency and cleanup in the discovery package - verifying no duplicate kinds or channel replacement on repeated AppendToMap calls, and that RemoveFromMap closes channels so reflectors stop cleanly.

Test Results:

`TestMemoryLeakSimulation` - 5 GVKs × 500 poll cycles

|                      | Buggy (pre-fix) | Fixed (post-fix) |
|----------------------|----------------:|----------------:|
| Kind entries in map  | 2500            | 5               |
| Stop channels live   | 5               | 5               |
| Heap growth (KB)     | +88             | -8              |

`TestGoroutineLeakSimulation` - 5 GVKs × 20 store rebuilds

|                   | Buggy (pre-fix) | Fixed (post-fix) |
|-------------------|----------------:|----------------:|
| Goroutines leaked | 100             | 0               |

Fixes #2867

### CODE REVIEW

File: internal/store/builder.go [hunk 1]
```diff
@@ -622,7 +622,17 @@ func (b *Builder) startReflector(
 	instrumentedListWatch := watch.NewInstrumentedListerWatcher(listWatcher, b.listWatchMetrics, reflect.TypeOf(expectedType).String(), useAPIServerCache, objectLimit, client)
 	reflector := cache.NewReflectorWithOptions(sharding.NewShardedListWatch(b.shard, b.totalShards, instrumentedListWatch), expectedType, store, cache.ReflectorOptions{ResyncPeriod: 0})
 	if cr, ok := expectedType.(*unstructured.Unstructured); ok {
-		go reflector.Run((*b.GVKToReflectorStopChanMap)[cr.GroupVersionKind().String()])
+		stopCh := make(chan struct{})
+		gvkStopCh := (*b.GVKToReflectorStopChanMap)[cr.GroupVersionKind().String()]
+		ctxDone := b.ctx.Done()
```
[the remark this answers is not in the document]
Comment (bhope) [reply internal/store/builder.go:627]: Fixed. The builder never touches the map directly now.

File: internal/store/builder_test.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2022 The Kubernetes Authors All rights reserved.
+Copyright 2026 The Kubernetes Authors All rights reserved.
```
[the remark this answers is not in the document]
Comment (bhope) [reply internal/store/builder_test.go:3]: Fixed, reverted back to the original year.

File: internal/store/builder_test.go [hunk 3]
```diff
@@ -258,3 +261,88 @@ func TestWithEnabledResources(t *testing.T) {
...
+// stopCh used for CR reflectors closes when the context is cancelled.
+func TestCRReflectorStopChanRespondsToContextCancel(t *testing.T) {
+	startGoroutines := runtime.NumGoroutine()
+
+	const n = 20
+	stopChs := make([]chan struct{}, n)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
```
[the remark this answers is not in the document]
Comment (bhope) [reply internal/store/builder_test.go]: Fixed. Removed the trailing time.Sleep + global goroutine count check.

File: internal/discovery/memleak_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	if fixedChannelCount != numGVKs {
+		t.Errorf("[fixed] expected exactly %d stop channels (stable), got %d", numGVKs, fixedChannelCount)
+	}
+
+	buggyGrowth := int64(heapAfterBuggy) - int64(heapBefore)      //nolint:gosec
+	fixedGrowth := int64(heapAfterFixed) - int64(heapBeforeFixed) //nolint:gosec
+	if buggyGrowth > 0 && fixedGrowth >= buggyGrowth {
+		t.Errorf("fixed heap growth (%d KB) should be less than buggy growth (%d KB)", fixedGrowth, buggyGrowth)
+	}
```
[the remark this answers is not in the document]
Comment (bhope) [reply internal/discovery/memleak_test.go]: Fixed. Applied the suggestion directly

File: internal/discovery/memleak_test.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+			go func(ch chan struct{}) {
+				<-ch
+			}(stopCh)
+		}
+	}
+
+	time.Sleep(20 * time.Millisecond)
+	goroutinesAfterBuggy := runtime.NumGoroutine()
+	buggyLeak := goroutinesAfterBuggy - goroutinesBefore
+
```
[the remark this answers is not in the document]
Comment (bhope) [reply internal/discovery/memleak_test.go]: Fixed. Rewrote TestGoroutineLeakSimulation to use a sync.WaitGroup

**Comment by bhope** [comment]
Hi @mrueg addressed the copilot suggestions and CI is now green. Ready for a review when you get a chance.

**Comment by jullianow** [comment]
Any idea when this will be released?

**Comment by bhope** [comment]
@jullianow This will be included in the upcoming release, we are working towards it. Please stay tuned. Thanks.

**Comment by mrueg** [comment]
> Hi @mrueg addressed the copilot suggestions and CI is now green. Ready for a review when you get a chance.


Thanks for looking into those comments. 
Unfortunately I won't have access to a way to test it until mid May due to private travel.

@rexagod can you take a look?

File: internal/store/builder_test.go
Comment (rexagod) [inline internal/store/builder_test.go:1]: The tests introduced here copy the behavior as it is now, and test it. However, since the behavior is not encapsulated (as a testable function), this needs to be updated every time there's a change. I'll suggest doing so and calling that here, or dropping these otherwise.

File: internal/discovery/memleak_test.go
Comment (rexagod) [inline internal/discovery/memleak_test.go:1]: No need to reproduce the current faulty behavior. It'd be fine to just see if the patched behavior works as expected. Tests implore maintenance too, so we'd benefit from testing traits in a way that balances manageable maintenance (PTAL at my other comment regarding reuse) with reasonable coverage, and dropping any extraneous ones that are arguably redundant in the long-term.

**Comment by alexandernorth** [comment]
Hi all, I took a look at this fix and I believe that the issue with the hanging channels would also be fixed by my PR #2872. However, I see that my PR is missing the parent context cancellation logic from builder.go.
Would you also be open to the approach I use regarding storing the `stopChan`s, but then adding the logic from this PR's builder.go to ensure that parent context cancellation also propagates correctly?
Let me know your thoughts!

**Comment by bhope** [comment]
Hi @alexandernorth - Thanks for the heads up. Since we already have done a few rounds of review on this one and left with a few test comments, I'd suggest we get this one merged first. Post that, I am happy to review your PR and get you the help needed to move forward on that one. Hope that works!

File: internal/store/builder_test.go
Comment (bhope) [reply internal/store/builder_test.go:1]: Thanks, makes sense. Extracted the bridge goroutine into newCRReflectorStopCh() in builder.go. Both tests now call that directly instead of copying the logic inside of the tests.

File: internal/discovery/memleak_test.go
Comment (bhope) [reply internal/discovery/memleak_test.go:1]: Thanks. Dropped:
1. appendToMapBuggy -     as you called out, no need to test the existence of bug
2. TestGoroutineLeakSimulation - it was duplicating the bridge goroutine logic from the store package and is already covered by the builder_test.go tests

Also, replaced TestMemoryLeakSimulation with TestAppendToMapStability which only asserts fixed behavior.

**Comment by bhope** [comment]
Hi @rexagod - addressed both your inline comments. PTAL when you get a chance. Thanks!

File: pkg/app/server.go [hunk 6]
```diff
@@ -325,7 +325,7 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
 			CRDsCacheCountGauge:     crdsCacheCountGauge,
 		}
 		// storeBuilder starts reflectors for the discovered GVKs, and as such, should close them too.
```
[the remark this answers is not in the document]
Comment (rexagod) [reply pkg/app/server.go:327]: Unchanged in this PR, and fairly legible as is.

**Comment by rexagod** [comment]
/approve
/lgtm
/hold

For other maintainers to review, feel free to unhold.

**Comment by jullianow** [comment]
@bhope When is this expected to be released?

**Comment by bhope** [comment]
@jullianow very soon, we are actively preparing the release. Stay tuned.

**Comment by jullianow** [comment]
thanks you @bhope
